compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -13314,113 +13577,92 @@ OMR::Z::TreeEvaluator::primitiveArraycopyEvaluator(TR::Node* node, TR::CodeGener
       }
    else
       {
-      // Don't clobber evaluate byteLenNode again if it was already evaluated
-      if (!alreadyEvaluatedByteLenNode)
+      // We need to decide direction of array copy at runtime.
+      TR::LabelSymbol *forwardArrayCopyLabel = generateLabelSymbol(cg);
+      cursor = generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::getCmpRegOpCode(), node, byteSrcReg, byteDstReg, TR::InstOpCode::COND_BNL, forwardArrayCopyLabel, false);
+      iComment("if byteSrcPointer >= byteDstPointer then GoTo forwardArrayCopy");
+      
+      // If it is contant length bytes array copy then Internal Control Flow starts here.
+      if (isConstantByteLen)
          {
-         byteLenReg = cg->gprClobberEvaluate(byteLenNode);
+         generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, startICFLabel);
+         startICFLabel->setStartInternalControlFlow();
          }
 
-      TR::Register* strideRegister = cg->allocateRegister();
-
-      MemCpyVarLenTypedMacroOp vtOp(node, byteDstNode, byteSrcNode, cg, node->getArrayCopyElementType(), byteLenReg, byteLenNode, false, node->isForwardArrayCopy());
-
-      vtOp.generate(byteDstReg, byteSrcReg, strideRegister);
-      vtOp.cleanUpReg();
-
-      cg->stopUsingRegister(strideRegister);
+      TR::Register *checkBoundReg = srm->findOrCreateScratchRegister();
+      if (byteLenReg == NULL)
+         byteLenReg = cg->gprClobberEvaluate(byteLenNode);
+      cursor = generateRRRInstruction(cg, TR::InstOpCode::getAddThreeRegOpCode(), node, checkBoundReg, byteSrcReg, byteLenReg); iComment("lastElementToCopy=byteSrcPointer+lengthInBytes");
+      
+      TR::LabelSymbol *backwardArrayCopyLabel = generateLabelSymbol(cg);
+      cursor = generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::getCmpRegOpCode(), node, checkBoundReg, byteDstReg, TR::InstOpCode::COND_BH, backwardArrayCopyLabel, false);
+      iComment("lastElementToCopy > byteDstPointer then GoTo backwardArrayCopy");
+      srm->reclaimScratchRegister(checkBoundReg);
+      
+      cursor = generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, forwardArrayCopyLabel);iComment("Forward Array Copy Sequence");
+      TR::TreeEvaluator::forwardArrayCopySequenceGenerator(node, cg, byteSrcReg, byteDstReg, byteLenReg, byteLenNode, srm, mergeLabel);
+      cursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, mergeLabel); iComment("Jump to MergeLabel");
+      
+      cursor = generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, backwardArrayCopyLabel); iComment("Backward Array Copy Sequence");
+      deps = TR::TreeEvaluator::backwardArrayCopySequenceGenerator(node, cg, byteSrcReg, byteDstReg, byteLenReg, srm, mergeLabel);
+      }
+   cursor = generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, mergeLabel);
+   if (!(isConstantByteLen && node->isForwardArrayCopy() && byteLenNode->getConst<int64_t>() < 256))
```

Comment:
I'm not really following what we're doing here exactly. Can you explain? This code seems out of place.

Reply:
So in case we have a constant length array copy in forward direction with length  within MVC instruction range, we will not have an ICF. This condition avoids putting generating endICF label if we do not have ICF

Reply:
Is there any downside to just putting it in there in terms of the actual emitter binary? I would prefer to keep the code clean at the expense of 2 labels being generated.

Reply:
Code inside this if do not generate a label. It sets end internal control flow to merge label and also creates dependencies for ICF. Also setting EndInternalControl Flow without start of internal control flow will hit an assert.  And main reason I am doing all this in if is to avoid generating dependencies in case there is no ICF.